### PR TITLE
[go_router] Refactor internal classes and methods

### PR DESCRIPTION
**Pull request by johnpryan**
This moves some things around to make things easier to change. The GoRouteInformationParser now returns a RouteMatchList object by delegating the matching to a RouteMatcher object. Once the initial matches are found, the parser processes redirects using an abstract RouteRedirector function. After redirecting, the parsing is done and the RouteMatchList is provided to GoRouterDelegate through setNewRoutePath, where it stores the current RouteMatchList. 

Once the GoRouterDelegate updates its RouteMatchList state, it builds the widget tree by delegating to RouteBuilder. RouteBuilder is a new class that builds widget tree starting with the top-level Navigator all the way down to the active sub-route.

List of changes:
- Separate matching from redirection
- Add RouteRedirector typedef
- Add RouteMatcher class
- Add RouteBuilder class
- Add RouteConfiguration class
- Rename and reorganize internal classes and libraries
- Add todo comments

I didn't modify the tests too much, but this should make it possible to write more specific tests for matching, redirection, building, etc.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/go_router/example/lib/router_stream_refresh.dart [hunk 1]
```diff
@@ -84,7 +84,8 @@ class _AppState extends State<App> {
         return null;
       },
       // changes on the listenable will cause the router to refresh it's route
```
Comment (loic-sharma) [inline packages/go_router/example/lib/router_stream_refresh.dart:86]:
```suggestion
      // changes on the listenable will cause the router to refresh its route
```

File: packages/go_router/lib/src/misc/refresh_stream.dart [hunk 2]
```diff
@@ -22,12 +18,12 @@ import 'go_router.dart';
 /// );
```
Comment (loic-sharma) [inline packages/go_router/lib/src/misc/refresh_stream.dart:18]: Looks like the example on line 17 needs to be updated

File: packages/go_router/lib/src/misc/refresh_stream.dart [hunk 3]
```diff
@@ -22,12 +18,12 @@ import 'go_router.dart';
...
 /// {@end-tool}
-class GoRouterRefreshStream extends ChangeNotifier {
-  /// Creates a [GoRouterRefreshStream].
+class StreamListenable extends ChangeNotifier {
+  /// Creates a [StreamListenable].
   ///
   /// Every time the [stream] receives an event the [GoRouter] will refresh its
   /// current route.
-  GoRouterRefreshStream(Stream<dynamic> stream) {
+  StreamListenable(Stream<dynamic> stream) {
```
Comment (loic-sharma) [inline packages/go_router/lib/src/misc/refresh_stream.dart:26]: Isn't this a breaking change? What made us decide to rename this? Should we update the changelog to call this out?

File: packages/go_router/lib/src/typedefs.dart [hunk 4]
```diff
@@ -2,22 +2,14 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'package:flutter/widgets.dart';
+import 'package:flutter/material.dart';
```
Comment (loic-sharma) [inline packages/go_router/lib/src/typedefs.dart:5]: Is this change necessary?

```suggestion
import 'package:flutter/widgets.dart';
```

File: packages/go_router/lib/src/parser.dart [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        redirector = redirect;
+
+  /// The route configuration for the app.
+  RouteConfiguration configuration;
+
+  /// The route matcher
+  RouteMatcher matcher;
+
+  /// The route redirector
+  RouteRedirector redirector;
```
Comment (loic-sharma) [inline packages/go_router/lib/src/parser.dart:30]: Should `configuration`, `matcher`, and `redirector` be final?

File: packages/go_router/lib/src/redirection.dart [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+RouteMatchList redirect(RouteMatchList prevMatchList,
+    RouteConfiguration configuration, RouteMatcher matcher,
+    {Object? extra}) {
+  RouteMatchList matches;
+
+  // Store each redirect to detect loops
+  final List<RouteMatchList> redirects = <RouteMatchList>[prevMatchList];
+
+  // Adds the redirect to the list of redirects if it is valid.
+  bool redirected(RouteMatchList newRedirect) {
```
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:27]: IMO this function is harder to read because `redirect` is nested inside. What do you think of unnesting `redirect`, renaming it to something like `_add_redirect`, and adding parameters as needed instead of capturing values from the parent scope? Something like `_add_redirect(redirects, newMatch, prevMatchList.location);`

File: packages/go_router/lib/src/redirection.dart [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    );
+
+    // If the new location is null, keep the matches the same as before.
+    if (topRedirectLocation != null) {
+      final RouteMatchList newMatch = matcher.findMatch(topRedirectLocation);
+
+      if (redirected(newMatch)) {
+        continue;
+      } else {
+        matches = newMatch;
```
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:85]: This case can only be reached if `newMatch` is null - will this cause issues when `matches` is used on line 104?

File: packages/go_router/lib/src/misc/refresh_stream.dart (hunk 3 above)
Comment (johnpryan) [reply packages/go_router/lib/src/misc/refresh_stream.dart:26]: Good catch, I didn't intend to make breaking changes here - I'll change this back. 

I would like to deprecate it or at the very least add some tests. For example, what happens when the stream is cancelled? Does this class work with single-subscription streams, or just broadcast streams? This class isn't really related to routing either.

File: packages/go_router/lib/src/parser.dart [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                .toString()));
+      }
+
+      // Use [SynchronousFuture] so that the initial url is processed
+      // synchronously and remove unwanted initial animations on deep-linking
+      return SynchronousFuture<RouteMatchList>(matches);
+    } on RedirectionError catch (e) {
+      final Uri uri = e.location;
+      return SynchronousFuture<RouteMatchList>(_errorScreen(uri, e.message));
+    } on MatcherError catch (e) {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/lib/src/parser.dart:82]: This is working as intended - the redirector also uses the matcher to find the match. For example, if you redirect from `/a` to `/b`, the redirector needs to get the matches for `/b`. Adding a comment.

File: packages/go_router/lib/src/redirection.dart (hunk 6 above)
Comment (johnpryan) [reply packages/go_router/lib/src/redirection.dart:27]: Sounds good to me

File: packages/go_router/lib/src/state.dart [hunk 9]
```diff
@@ -4,13 +4,13 @@
...
 
-import 'go_route_information_parser.dart';
+import 'configuration.dart';
 
 /// The route state during routing.
 class GoRouterState {
   /// Default constructor for creating route state during routing.
   GoRouterState(
-    this._delegate, {
+    this._configuration, {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/lib/src/state.dart:13]: GoRouterState is somewhat misnamed... It's not really the "state" because it's not stored anywhere - it's constructed by the builder to provide the user with information. So I don't expect users to be calling the constructor. Maybe we should add a TODO to make it abstract so we can keep the constructor private...

File: packages/go_router/lib/go_router.dart [hunk 10]
```diff
@@ -6,24 +6,284 @@
...
+export 'src/pages/custom_transition_page.dart';
+export 'src/platform.dart' show UrlPathStrategy;
+export 'src/typed_routing.dart' show GoRouteData, TypedGoRoute;
+export 'src/typedefs.dart';
+
+/// The top-level go router class.
+///
+/// Create one of these to initialize your app's routing policy.
+// ignore: prefer_mixin
+class GoRouter extends ChangeNotifier with NavigatorObserver {
```
Comment (loic-sharma) [inline packages/go_router/lib/go_router.dart:30]: What's the benefit of moving `GoRouter` to this file? Is it to cut down the number of files, or are there additional benefits?

I'm concerned that this forces things that depend on `GoRouter` to also be embedded in this file (like `InheritedGoRouter`), which may cause this file to bloat over time.

File: packages/go_router/lib/src/configuration.dart [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  /// The limit for the number of consecutive redirects
+  final int redirectLimit;
+
+  /// Top level page redirect.
+  final GoRouterRedirect topRedirect;
+
+  final Map<String, String> _nameToPath = <String, String>{};
+
+  /// Throws a [RouteConfigurationError] if this configuration is invalid
+  void validate() {
```
Comment (loic-sharma) [inline packages/go_router/lib/src/configuration.dart:40]: Ideally one shouldn't be able to create an invalid `RouteConfiguration` instance. Are there any scenarios where you'd want to create a `RouteConfiguration` without running `validate`? If no, what do you think of making this a factory constructor?

Comment (loic-sharma) [reply packages/go_router/lib/src/configuration.dart:40]: Ah the parser used to enforce this validation in its constructor:

https://github.com/flutter/packages/blob/a637b4a038c69b079d55d5b8e07bc823a279e162/packages/go_router/lib/src/go_route_information_parser.dart#L32-L39

File: packages/go_router/lib/src/redirection.dart [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+  String toString() =>
+      super.toString() +
+      <String>[
+        ...matches.map(
+            (RouteMatchList routeMatches) => routeMatches.location.toString()),
+      ].join(' => ');
+}
+
+/// Adds the redirect to [redirects] if it is valid.
+/// Returns true if the redirect was processed.
```
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:122]:
```suggestion
```

File: packages/go_router/lib/src/redirection.dart [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    // If the new location is null, keep the matches the same as before.
+    if (topRedirectLocation != null) {
+      final RouteMatchList newMatch = matcher.findMatch(topRedirectLocation);
+
+      _addRedirect(redirects, newMatch, prevMatchList.location,
+          configuration.redirectLimit);
+      continue;
+    } else {
+      matches = currentMatches;
+    }
```
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:54]: Nit: what do you think of removing the `else` and moving the comment from line 45 down?

```suggestion
    if (topRedirectLocation != null) {
      final RouteMatchList newMatch = matcher.findMatch(topRedirectLocation);

      _addRedirect(redirects, newMatch, prevMatchList.location,
          configuration.redirectLimit);
      continue;
    }

    // If there's no top-level redirect, keep the matches the same as before.
    matches = currentMatches;
```

File: packages/go_router/lib/src/configuration.dart [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    return sb.toString();
+  }
+
+  void _debugFullPathsFor(
+      List<GoRoute> routes, String parentFullpath, int depth, StringBuffer sb) {
+    for (final GoRoute route in routes) {
+      final String fullpath = concatenatePaths(parentFullpath, route.path);
+      assert(() {
```
Comment (loic-sharma) [inline packages/go_router/lib/src/configuration.dart:111]: Is this `assert` intentional now that this is not logging directly?

File: packages/go_router/lib/src/configuration.dart [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    required this.topRedirect,
+  }) {
+    _cacheNameToPath('', routes);
+    assert(() {
+      _debugKnownRoutes();
+      return true;
+    }());
+  }
+
+  /// List of top level routes used by the go router delegate
```
Comment (loic-sharma) [inline packages/go_router/lib/src/configuration.dart:28]: It looks like periods were removed from the end of all comments when this was moved out of `GoRouteInformationParser`. Is that intentional? My understanding is the Flutter style guide is to always have periods at the end of comments.

File: packages/go_router/lib/src/delegate.dart [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+          configuration: configuration,
+          builderWithNav: builderWithNav,
+          errorPageBuilder: errorPageBuilder,
+          errorBuilder: errorBuilder,
+          restorationScopeId: restorationScopeId,
+          observers: observers,
+        );
+
+  /// Builds the top-level Navigator given a configuration and location.
+  final RouteBuilder builder;
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply packages/go_router/lib/src/delegate.dart:39]: (I ask because there's a member below with `@visibleForTesting`)

File: packages/go_router/lib/go_router.dart [hunk 17]
```diff
@@ -6,24 +6,284 @@
...
+  late final RouteConfiguration routeConfiguration;
+
+  /// The route information parser used by the go router.
+  late final GoRouterInformationParser routeInformationParser;
+
+  /// The router delegate used by the go router.
+  late final GoRouterDelegate routerDelegate;
+
+  /// The route information provider used by the go router.
+  late final GoRouteInformationProvider routeInformationProvider;
```
Comment (loic-sharma) [inline packages/go_router/lib/go_router.dart:110]: This issue is interesting: https://github.com/flutter/flutter/issues/106863

It seems these are our options:

1. Do nothing. Users need to import files manually to use these fields 🤮 
2. Export more files. Make `GoRouterInformationParser`, `GoRouterDelegate`, and `GoRouteInformationProvider` public. This makes it harder for us to make breaking changes in the future.
3. Make this expose the "base" types that are already public (like `RouteInformationProvider` instead of `GoRouteInformationProvider`). This is a breaking change and it may not solve all users' scenarios. For example, a user might need something from `GoRouteInformationProvider` that isn't available from `RouteInformationProvider`.

What are your thoughts on this? This can be solved in a future PR, but I'm curious on what direction you have in mind.

File: packages/go_router/lib/src/configuration.dart (hunk 11 above)
Comment (johnpryan) [reply packages/go_router/lib/src/configuration.dart]: I agree, how about we throw an error in the `RouteConfiguration` constructor? https://github.com/flutter/packages/pull/2317/commits/db91f560bcc39434a5320b8b3e6a98c6f2c83ad1

File: packages/go_router/lib/src/redirection.dart (hunk 13 above)
Comment (johnpryan) [reply packages/go_router/lib/src/redirection.dart:54]: SGTM

File: packages/go_router/lib/src/delegate.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/lib/src/delegate.dart:39]: Yes, good catch, in fact, how do you feel about adding this to all of its methods? For example, the intended use is `GoRouter.of(context).push()` rather than `GoRouter.of(context).routerDelegate.push()`.

File: packages/go_router/lib/src/configuration.dart (hunk 14 above)
Comment (johnpryan) [reply packages/go_router/lib/src/configuration.dart:109]: I'm not sure why this named location feature needs so many `asserts`.  I'd like to simplify how `namedLocation` works or remove the API altogether.

File: packages/go_router/lib/src/configuration.dart (hunk 15 above)
Comment (johnpryan) [reply packages/go_router/lib/src/configuration.dart:36]: SGTM - I went ahead and added periods to all of the doc comments.

File: packages/go_router/lib/go_router.dart (hunk 17 above)
Comment (johnpryan) [reply packages/go_router/lib/go_router.dart]: I left a [comment](https://github.com/flutter/flutter/issues/106863#issuecomment-1184739862) to see if there's a reason these need to be assigned to variables.

My first thought was to change the type from `GoRouteInformationParser` to `RouteInformationParser` but this produces a warning because we have the [always_specify_types](https://dart-lang.github.io/linter/lints/always_specify_types.html) lint enabled. So I think we should put `@visibleForTesting` on these classes and encourage users to use the methods defined on GoRouter instead. We could also add an `// ignore: always_specify_types` comment here.

Comment (atoka93) [reply packages/go_router/lib/go_router.dart]: Answered it here: https://github.com/flutter/flutter/issues/106863#issuecomment-1185291219
Since I noticed this: if it would be a vote I would prefer option 3 over 2, but at the end fine with either one.

File: packages/go_router/lib/src/delegate.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (loic-sharma) [reply packages/go_router/lib/src/delegate.dart:40]: That sounds good to me 👍

File: packages/go_router/lib/go_router.dart (hunk 17 above)
Comment (johnpryan) [reply packages/go_router/lib/go_router.dart]: I filed https://github.com/flutter/flutter/issues/107757, which I'd like to include in the next major version (5.0.0)

File: packages/go_router/lib/src/delegate.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/lib/src/delegate.dart:40]: I'll do this as part of https://github.com/flutter/flutter/issues/107757

File: packages/go_router/lib/src/builder.dart [hunk 18]
```diff
@@ -151,18 +104,19 @@ class GoRouterDelegate extends RouterDelegate<List<GoRouteMatch>>
...
     final Uri uri = Uri.parse(location);
     return builderWithNav(
       context,
       GoRouterState(
-        _parser,
+        configuration,
         location: location,
-        name: null, // no name available at the top level
+        name: null,
+        // no name available at the top level
```
Comment (loic-sharma) [inline packages/go_router/lib/src/builder.dart:119]: What do you think of moving comments to before the argument? Same comment for line 130.

```suggestion
        // no name available at the top level
        name: null,
```

File: packages/go_router/lib/src/parser.dart [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+  Future<RouteMatchList> parseRouteInformation(
+    RouteInformation routeInformation,
+  ) {
+    assert(() {
+      if (debugRequireGoRouteInformationProvider) {
+        assert(
+            routeInformation is DebugGoRouteInformation,
+            'This GoRouteInformationParser needs to be used with '
+            'GoRouteInformationProvider, did you forget to pass in '
+            'GoRouter.routeInformationProvider to the Router constructor?');
```
Comment (loic-sharma) [inline packages/go_router/lib/src/parser.dart:54]: Does this use the same style guide as the flutter repo? If so, I think this should be like this:

```suggestion
        assert(
            routeInformation is DebugGoRouteInformation,
            'This GoRouteInformationParser needs to be used with '
            'GoRouteInformationProvider, did you forget to pass in '
            'GoRouter.routeInformationProvider to the Router constructor?',
        );
```

**Review by loic-sharma — APPROVED** [review]
LGTM but please make sure to check API docs for unexpected breaking changes before merging!

**Comment by johnpryan** [comment]
I diffed the API docs this branch and `main`, and found that we were only exporting two typedefs, so I changed that back: https://github.com/flutter/packages/pull/2317/commits/7f8954c7a69ee245f0042cdfbec7f2efa31e003b. Other than that, I don't think this will be a breaking change.

Technically, we are also exporting `routeConfiguration` on `GoRouter`, but it is marked `@visibleForTesting` so I don't consider this a reason to bump the minor version

File: packages/go_router/lib/src/builder.dart [hunk 20]
```diff
@@ -45,77 +36,39 @@ class GoRouterDelegate extends RouterDelegate<List<GoRouteMatch>>
...
-
-  /// For use by the Router architecture as part of the RouterDelegate.
-  @override
-  Future<void> setNewRoutePath(List<GoRouteMatch> configuration) {
-    _matches = configuration;
-    // Use [SynchronousFuture] so that the initial url is processed
-    // synchronously and remove unwanted initial animations on deep-linking
-    return SynchronousFuture<void>(null);
-  }
+  /// NavigatorObserver used to receive change notifications when navigation
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:46]:
```suggestion
  /// NavigatorObserver used to receive notifications when navigating in between routes.
```

File: packages/go_router/lib/src/configuration.dart [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+import 'configuration.dart';
+import 'logging.dart';
+import 'path_utils.dart';
+import 'typedefs.dart';
+
+export 'route.dart';
+export 'state.dart';
+
+/// The route configuration for GoRouter configured by the app.
+class RouteConfiguration {
```
Comment (chunhtai) [inline packages/go_router/lib/src/configuration.dart:14]: The properties in this class are only needed in gorouteinformationparser once we remove the namedLocation. I am a bit unsure whether we should introduce a new class that will probably be deprecated in the future

File: packages/go_router/lib/src/information_provider.dart [hunk 22]
```diff
@@ -111,9 +110,9 @@ class GoRouteInformationProvider extends RouteInformationProvider
 }
 
 /// A debug class that is used for asserting the [GoRouteInformationProvider] is
-/// in use with the [GoRouteInformationParser].
+/// in use with the [RouteInformationParser].
```
Comment (chunhtai) [inline packages/go_router/lib/src/information_provider.dart:113]: this meant to specifically mention GoRouteInformationParser, can you revert this change?

File: packages/go_router/lib/src/match.dart [hunk 23]
```diff
@@ -4,16 +4,15 @@
...
-class GoRouteMatch {
-  /// Constructor for GoRouteMatch, each instance represents an instance of a
-  /// GoRoute for a specific portion of a location.
-  GoRouteMatch({
+import 'path_utils.dart';
+import 'route.dart';
+
+/// Each RouteMatch instance represents an instance of a GoRoute for a specific
+/// portion of a location.
+class RouteMatch {
```
Comment (chunhtai) [inline packages/go_router/lib/src/match.dart:12]: I am curious why we rename this class while most of the other classes still have GoRoute in it.

File: packages/go_router/lib/src/builder.dart [hunk 24]
```diff
@@ -151,18 +104,19 @@ class GoRouterDelegate extends RouterDelegate<List<GoRouteMatch>>
 
     // pass either the match error or the build error along to the navigator
     // builder, preferring the match error
-    if (matches.length == 1 && matches.first.error != null) {
-      error = matches.first.error;
+    if (matches.matches.length == 1 && matches.matches.first.error != null) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:107]: maybe this can be a boolean property or a getter in RouteMatchList?

File: packages/go_router/lib/src/router.dart [hunk 25]
```diff
@@ -26,16 +23,19 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
...
     required List<GoRoute> routes,
+    // TODO(johnpryan): Change to a route, improve error API
     GoRouterPageBuilder? errorPageBuilder,
     GoRouterWidgetBuilder? errorBuilder,
     GoRouterRedirect? redirect,
     Listenable? refreshListenable,
     int redirectLimit = 5,
     bool routerNeglect = false,
     String? initialLocation,
+    // TODO(johnpryan): Deprecate this parameter
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:34]: I can imagine people would want to control their path strategy, are we going to replace it with a different api?

File: packages/go_router/lib/src/router.dart [hunk 26]
```diff
@@ -82,26 +87,44 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
...
+
+  /// The route information provider used by [GoRouter].
+  // TODO(johnpryan): change type to RouteInformationProvider
+  GoRouteInformationProvider get routeInformationProvider =>
+      _routeInformationProvider;
 
-  /// The router delegate used by the go router.
-  late final GoRouterDelegate routerDelegate;
+  /// The route information parser used by [GoRouter].
+  // TODO(johnpryan): change type to RouteInformationParser
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:106]: I feel we will still need to expose the GoRoute- delegates if customers want to create their own delegates to use with the GoRouter- delegates

**Review by chunhtai — COMMENTED** [review]
I like the direction this pr is going, but I am a bit concern about some of the class renaming and this can be a breaking change for developer who use the goroute- delegates directly.

File: packages/go_router/example/lib/router_stream_refresh.dart [hunk 27]
```diff
@@ -84,6 +84,7 @@ class _AppState extends State<App> {
...
+      // TODO(johnpryan): Deprecate GoRouterRefreshStream
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/example/lib/router_stream_refresh.dart:87]: Here's my thinking: https://github.com/flutter/flutter/issues/108128

File: packages/go_router/lib/src/configuration.dart (hunk 21 above)
Comment (johnpryan) [reply packages/go_router/lib/src/configuration.dart:14]: The routeConfiguration field is marked `@visibleForTesting` in RouterDelegate, so I don't consider this a new public API. The reason for making this a separate class is make it easier to change the configuration without updating all of the places where the routeConfiguration is being passed around. For example, we might want to deprecate `topRedirect`.

File: packages/go_router/lib/src/information_provider.dart (hunk 22 above)
Comment (johnpryan) [reply packages/go_router/lib/src/information_provider.dart:113]: good catch - fixed

File: packages/go_router/lib/src/match.dart (hunk 23 above)
Comment (johnpryan) [reply packages/go_router/lib/src/match.dart:12]: This class isn't part of the public API, so we can safely rename without bumping the major version yet.

File: packages/go_router/lib/src/configuration.dart (hunk 21 above)
Comment (chunhtai) [reply packages/go_router/lib/src/configuration.dart:14]: but at somepoint we won't need to pass it around.

We would need to expose this class if we expose the GoRouterDelegate, otherwise, developer won't be able to use GoRouterDelegate constructor directly.

File: packages/go_router/example/lib/router_stream_refresh.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router/example/lib/router_stream_refresh.dart:87]: gotcha and agree this should not belong to gorouter

File: packages/go_router/lib/src/builder.dart (hunk 24 above)
Comment (johnpryan) [reply packages/go_router/lib/src/builder.dart:107]: Yeah that would be better - added.

File: packages/go_router/lib/src/router.dart (hunk 25 above)
Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:34]: The `flutter_web_plugins` package (part of the SDK) now takes care of the conditional imports for you, so using this API is simpler:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_web_plugins/url_strategy.dart';

void main() {
  usePathUrlStrategy();
  runApp(ExampleApp());
}
```

See also: https://github.com/flutter/website/issues/7386

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:34]: sounds good, can you add the issue link to the todo? same thing as other TODO that may have issue associate with it

File: packages/go_router/lib/src/match.dart (hunk 23 above)
Comment (chunhtai) [reply packages/go_router/lib/src/match.dart:12]: my question was more of why we rename it.

It was probably my mistake to not make GoRouteMatch public because GoRouteInformationParser already expose GoRouteMatch as generic type

File: packages/go_router/lib/src/router.dart (hunk 26 above)
Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:106]: This class isn't currently part of the public API, but yes if we wanted to make it  configurable as the part of the constructor we would have to export it:


```dart
import 'package:go_router/go_router.dart'
class CustomMatchListParser extends MatchListParser {
  // ...
}

// ... 
  final _router = GoRouter(
    builder: CustomMatchListBuilder(),
    parser: CustomMatchListParser(),
    matcher: CustomMatchListMatcher(),
    redirector: CustomMatchListRedirector(),
    routes: [
      // ...
    ],
```

We could take it one step further and allow users to specify a custom implementation of RouteConfiguration.

File: packages/go_router/lib/src/router.dart (hunk 25 above)
Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:34]: Filed an issue: https://github.com/flutter/flutter/issues/108132

File: packages/go_router/lib/src/router.dart (hunk 26 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:106]: I mean unless we are never going to expose GoRouteInfromationParser/GoRouterDelegate, it doesn't make sense to change the type back to RouteInformationParser given that it is already exposed as GoRouteInformationParser. I think right now this package is in a weird state that some classes should have been exposed as public but don't, and it was too late to making it completely private because some other classes have exposed them as part of their APIs

Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:106]: I think I misunderstood your question, sorry. There was a discussion about this in a previous review: https://github.com/flutter/packages/pull/2317#discussion_r921422617

File: packages/go_router/lib/src/match.dart (hunk 23 above)
Comment (johnpryan) [reply packages/go_router/lib/src/match.dart:12]: I didn't see a reason to put `Go` in front of internal classes like RouteMatcher, RouteMatch, etc.

File: packages/go_router/lib/src/configuration.dart (hunk 21 above)
Comment (johnpryan) [reply packages/go_router/lib/src/configuration.dart:14]: Filed an issue to remove this along with the namedLocation API: https://github.com/flutter/flutter/issues/108139

File: packages/go_router/lib/src/router.dart (hunk 25 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:34]: can you link this issue in the todo here?

File: packages/go_router/lib/src/match.dart (hunk 23 above)
Comment (chunhtai) [reply packages/go_router/lib/src/match.dart:12]: sounds good

File: packages/go_router/example/lib/router_stream_refresh.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router/example/lib/router_stream_refresh.dart:87]: can you link the issue in the todo here?

File: packages/go_router/lib/src/router.dart (hunk 25 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:34]: here and everywhere

Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:35]: Done

File: packages/go_router/example/lib/router_stream_refresh.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (johnpryan) [reply packages/go_router/example/lib/router_stream_refresh.dart:87]: Done: https://github.com/flutter/packages/pull/2317/commits/07f7ada1834413273e1ac2d5ebafefbb8fdf8d6d

File: packages/go_router/lib/src/state.dart [hunk 28]
```diff
@@ -29,8 +29,8 @@ class GoRouterState {
                     : subloc),
         assert((path ?? '').isEmpty == (fullpath ?? '').isEmpty);
 
-  // TODO(chunhtai): remove this once namedLocation is removed from go_router.
-  final GoRouteInformationParser _delegate;
+  // TODO(johnpryan): remove once namedLocation is removed from go_router.
```
Comment (chunhtai) [inline packages/go_router/lib/src/state.dart:32]: Maybe link this ? https://github.com/flutter/flutter/issues/107729

File: packages/go_router/lib/src/state.dart [hunk 29]
```diff
@@ -64,12 +64,13 @@ class GoRouterState {
 
   /// Get a location from route name and parameters.
   /// This is useful for redirecting to a named location.
+  // TODO(johnpryan): deprecate namedLocation API
```
Comment (chunhtai) [inline packages/go_router/lib/src/state.dart:67]: same https://github.com/flutter/flutter/issues/107729

**Review by chunhtai — APPROVED** [review]
LGTM
